dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackerOwnershipManager.java

```diff
@@ -254,15 +253,20 @@ public boolean hasAccess( User user, TrackedEntityInstance entityInstance, Progr
             return true;
         }
 
-        OrganisationUnit ou = getOwner( entityInstance.getId(), program, entityInstance::getOrganisationUnit );
+        OrganisationUnit ownerOrgUnit = getOwner( entityInstance.getId(), program );
+        boolean hasOwner = ownerOrgUnit != null;
 
-        if ( program.isOpen() || program.isAudited() )
+        if ( !hasOwner )
         {
-            return organisationUnitService.isInUserSearchHierarchyCached( user, ou );
+            return organisationUnitService.isInUserSearchHierarchyCached( user, entityInstance.getOrganisationUnit() );
```

Comment:
I think we go to search scope if program is open or audited, otherwise we restrict to capture scope.

Reply:
The discussion we had in the design meeting, was about creating an enrollment in a CLOSED program, when there is no ownership within the context of that program.

As we agreed, ownership only makes sense in the case of a program. If no enrollment exists for a program, ownership goes out of the window (Since we don't have ownership in the context of that program). In that case we fall back to the TEI ou.

In all cases (Let me know if I am wrongly assuming here), the user needs CAPTURE scope access for a OU to create any data there, including an enrollment.

In this case:
TEI.ou = A
Enrollment.ou = B

A needs to be in the search scope (We need to be able to see the TEI)
B needs to be in the capture scope (We need "capture access" to the ou where we create the enrollment).


In the case over, we determine if a user has access to a tei through the context of a program. In our usecase, the tei don't have an owner for the provided program, so we fall back to the tei ou - Which means, the tei.ou needs to be in the search scope for the user to have access.

An additional check is required in another place to make sure the Enrollment.ou is actually in the users capture scope.

Both of these are regardless of access level. The access level is only relevant for reading, or for updating existing data that has an owner.

To me, the code over seems correct to me. It's the same as before (logic) with the exception that we now check if ownership exists or not, and create this specific exception

Reply:
Thank you @stian-sandvold for your detailed explanation 😄 So statement

> If a program has no owner, use search scope instead of capture scope to allow registering a TEI to the program.

is only related to this check

> TEI.ou = A: A needs to be in the search scope (We need to be able to see the TEI)

And

> In all cases (Let me know if I am wrongly assuming here), the user needs CAPTURE scope access for a OU to create any data there, including an enrollment.

is checked somewhere else. Did I get that right?

And is the definition of a closed program one that has no owner? Or one in which you cannot enroll?